### ISSUE TITLE
[Phase II] rl_alpha — task3

### ISSUE THREAD

**Issue by game-loader** [EBiM-Benchmark/submissions#34]
### Team name

rl_alpha

### Point of Contact email

983028860@qq.com

### Task

Task 3 — Assisted Living & Feeding

### Your assigned Phase II pathway for this task

Shanghai

### Public GitHub repository URL

https://github.com/game-loader/ebim_phase2_task3

### Pinned commit SHA

d73fe482144410098b822f79ac84bbdff3aac840

## Build (arm host, or any x86-64 host with Docker; no GPU needed)

  ```bash
  git clone git@github.com:game-loader/ebim_phase2_task3.git
  cd ebim_phase2_task3
  git checkout d73fe482144410098b822f79ac84bbdff3aac840
  docker build -t franka-duo-table-mission:phase2 .
  ```

  The build runs 66 offline tests as its last layer; a successful build proves the image is complete.

  ## One-time deployment of our host-side packages (none exist on a stock machine)

  ```bash
  # Arm host (ROS 2 Jazzy): our controller overlay incl. the JointImpedanceController, env, DDS config
  rsync -a hosts/arm/teleoperation_overlay/ <arm-user>@<arm-host>:~/recloned_sources/teleoperation_overlay/
  rsync -a hosts/arm/env/tmr_env.sh hosts/arm/env/cyclonedds.xml <arm-user>@<arm-host>:~/
  rsync -a hosts/arm/env/source_migrated_stack.sh <arm-user>@<arm-host>:~/recloned_sources/
  ssh <arm-user>@<arm-host> 'cd ~/recloned_sources/teleoperation_overlay && source
  ~/recloned_sources/franka_ros2_jazzy_ws/install/setup.bash && colcon build --symlink-install'

  # Arm host: our joint servo (from this checkout, on the arm host)
  source ~/tmr_env.sh
  colcon build --base-paths site --build-base site/build --install-base site/install --merge-install \
    --packages-select franka_duo_joint_servo franka_duo_ptp_step

  # Base host (ROS 2 Humble): our navigation routes, adapter package, home scripts, vendor patches
  rsync -a base/tmr_cycle/      <base-user>@<base-host>:~/tmr_cycle/
  rsync -a base/tmr_navigation/ <base-user>@<base-host>:~/tmr_navigation/
  rsync -a hosts/base/home/     <base-user>@<base-host>:~/
  ssh <base-user>@<base-host> 'cd ~/ros2_ws/src/franka_ros2 && git apply
  ~/vendor_patches/franka_bringup_tmr_config.patch; \
    cd ~/ros2_ws/src/zed-ros2-wrapper && git apply ~/vendor_patches/zed_ros2_wrapper.patch; \
    cd ~/tmr_navigation && source /opt/ros/humble/setup.bash && colcon build --symlink-install'
  ```

  ## Bring-up before each run

  ```bash
  # Base host
  cd ~/tmr_cycle && bash scripts/19_ensure_navigation_stack.sh && bash scripts/17_control_mode.sh mission

  # Arm host (source ~/tmr_env.sh first)
  ros2 launch franka_fr3_arm_controllers franka.launch.py arm_id:=fr3v2 arm_prefix:=left  namespace:=left
  robot_ip:=<LEFT_IP>  load_gripper:=false joint_sources:=joint_states
  ros2 launch franka_fr3_arm_controllers franka.launch.py arm_id:=fr3v2 arm_prefix:=right namespace:=right
  robot_ip:=<RIGHT_IP> load_gripper:=false joint_sources:=joint_states
  ros2 launch franka_gripper_manager robotiq_gripper_controller_client.launch.py
  config_file:=example_fr3_duo_config_robotiq.yaml
  ros2 launch franka_spine_server spine.launch.py spine_ip:=<SPINE_IP>
  ros2 launch zed_wrapper zed_camera.launch.py camera_model:=zedm camera_name:=head_camera publish_tf:=false
  ros_params_override_path:=~/zed_override.yaml
  source site/install/setup.bash && bash scripts/start_servo_and_activate.sh 0.1 --reconfigure   # wait for "READY:"
  ```

  ## Run the policy (arm host)

  ```bash
  # Dry run: prints the 12-step plan, moves nothing.
  docker run --rm --network host franka-duo-table-mission:phase2 mission \
    --base-host <base-user>@<base-host> --base-root /home/<base-user>/tmr_cycle

  # Full mission: drives the base and both arms. Entrypoint: /app/entrypoint.sh mission
  docker run --rm --network host \
    -v ~/.ssh:/root/.ssh:ro \
    -v "$PWD/site/install:/app/site/install:ro" \
    -v "$PWD/outputs:/app/outputs" \
    franka-duo-table-mission:phase2 mission \
      --base-host <base-user>@<base-host> --base-root /home/<base-user>/tmr_cycle \
      --speed 0.1 --execute
  ```

## Container

  - **Base image:** `ros:jazzy-ros-base` (Ubuntu 24.04, ROS 2 Jazzy).
  - **CUDA / GPU:** none. Perception is `yolo11n-seg` on CPU (~15 ms/frame after warm-up); CPU-only PyTorch wheels are
  installed. No NVIDIA driver or `--gpus` flag needed.
  - **Python deps (in-image venv, `--system-site-packages` so apt `rclpy` stays importable):** numpy 1.23–2.2, PyYAML
  6, opencv-python-headless 4.9+, ultralytics 8.3.x (with CPU torch), pytest.
  - **ROS deps (apt, in image):** `ros-jazzy-{sensor,geometry,std,nav,control,controller-manager}-msgs`,
  `ros-jazzy-rmw-cyclonedds-cpp`. `RMW_IMPLEMENTATION=rmw_cyclonedds_cpp`, `ROS_DOMAIN_ID=0` by default; override with
  `-e` if the arm host differs.
  - **Baked into the image (no downloads):** YOLO weights `assets/weights/yolo11n-seg.pt` (6.2 MB), head-camera
  extrinsics `configs/zed_pnp_calibration.json`, 20-D action contract `assets/contract/`, taught stow posture
  `configs/grasp_stage_poses.json`. Ultralytics is forced offline (`YOLO_OFFLINE=1`).

  ## Network at run time

  **No internet access is needed or used.** The container does require:
  - `--network host` — the policy is a ROS 2 (CycloneDDS) participant that must discover the arm host's driver nodes.
  - SSH from the container to the base host (mounted key, `-v ~/.ssh:/root/.ssh:ro`) to launch the navigation routes.
  This is LAN-only.

  ## Must already be present on the host

  **Arm host (ROS 2 Jazzy):**
  - Franka `franka_ros2` (Jazzy) workspace incl. `franka_hardware`, `franka_robot_state_broadcaster`,
  `franka_spine_server`, `franka_description`; `libfranka` (`95b406f4`).
  - **Our overlay `hosts/arm/teleoperation_overlay/`** built on top of it — this is where the
  `JointImpedanceController` (`franka_fr3_arm_controllers`), Robotiq `franka_gripper_manager`, and `franka_spine_msgs`
  come from. Not on a stock machine.
  - **Our joint servo `site/franka_duo_joint_servo`** built with colcon from this checkout; its `install/` is
  bind-mounted into the container.
  - ZED SDK + `zed-ros2-wrapper` (ZED Mini, serial 17064700), publishing rectified RGB 640×360 + `camera_info` (depth
  disabled).
  - Running at launch time: both FR3 drivers (`/left`, `/right`), both Robotiq drivers, ZED wrapper,
  `franka_spine_node`, joint servo + gello relay with both `joint_impedance_controller`s active
  (`scripts/start_servo_and_activate.sh 0.1`).
  - `~/tmr_env.sh` + `~/cyclonedds.xml` from `hosts/arm/env/`.

  **Base host (ROS 2 Humble, separate DDS domain 97):**
  - `franka_ros2` humble (`1006036`) with `franka_bringup` (swerve base) + `franka_mobile_sensors`;
  `sick_safetyscanners2` (`886a81a`) for the two LiDARs; `slam_toolbox`.
  - **Our `base/tmr_cycle/`** (routes 07/13/15/20, velocity adapter, live table-leg detection) and **our
  `base/tmr_navigation/`** (odom adapter, dual-LiDAR merger) — not on a stock machine; deploy per README §2.1.
  - Two small vendor patches applied (`hosts/base/vendor_patches/`): `franka_bringup` `use_rviz: false`;
  `zed-ros2-wrapper` 30 Hz / `depth_mode: NONE`.
  - Running at launch time: base controller, odometry, both LiDARs, SLAM, exactly one `cmd_vel_adapter.py` in mission
  mode (`scripts/19_ensure_navigation_stack.sh` then `17_control_mode.sh mission`).
  - Passwordless SSH login from the arm host.

  **Docker:** any recent Docker Engine; no GPU runtime.

  Pinned vendor revisions for every upstream package are tabulated in README §2.1.

## Compute

  - **GPU: none required.** Perception (`yolo11n-seg`) runs on CPU, ~15 ms/frame after warm-up. No VRAM requirement.
  - **Rates:** policy emits 20-D dual-arm Cartesian action chunks at `30 Hz × playback_speed`, default `0.1` → 3 Hz
  effective. Perception is one frame per table stage (not continuous). The 1 kHz joint tracking (Ruckig) and 1 kHz
  joint impedance loops run in host-side ROS 2 nodes on the arm host, not in the container. Base navigation runs at 20
  Hz on the base host.

  ## Platform

  - Franka Duo Mobile: 2× FR3 (`fr3v2`) on the `franka_spine_v0_1` lifting spine, TMR swerve base, 2× Robotiq 2F-85,
  ZED Mini head camera, 2× SICK LiDAR (front/rear). Wrist cameras are not used.
  - Both arms are mounted on the spine carriage, so spine height moves both arms. The policy raises the spine to
  **0.700 m** before every drive and lowers it to **0.468 m** at every table.

  ## Camera placement

  - ZED Mini fixed on the head mount, rectified RGB **640×360**, depth disabled. Extrinsics
  (`configs/zed_pnp_calibration.json`) were solved by hand-eye PnP **at spine 0.468 m** and are only valid at that
  height and for this mount. A camera moved on its mount, or a different grasp height, requires re-calibration
  (procedure in `docs/ZED_PNP_CALIBRATION.md`).
  - Perception assumes the whole tray is inside the head-camera field of view when the base has stopped at the pick
  table (base route 07 stops ~0.7 m past the doorway, facing the table).

  ## Initial arm pose

  - Both arms in the shipped `travel_stow` posture (`configs/grasp_stage_poses.json`, 18-D EE pose in the arm-base
  midpoint frame) — folded inside the base collision envelope (front/rear 0.40 m, width 0.58 m). The mission's first
  phase drives the arms there itself via the impedance controller, so any start pose reachable without collision is
  acceptable; grippers **open**.
  - Both `joint_impedance_controller`s active, joint servo idle-latched (`start_servo_and_activate.sh` prints
  `READY:`). Spine height at start is irrelevant; it is commanded to 0.700 m before departure.

  ## Base start

  - Base at the marked start pose and heading (route 07 is odometry-relative: forward 0.85 m → CW 90° → live doorway
  detection → through the door). Doorway and table legs are detected live from the dual LiDARs; no map coordinates are
  assumed.

  ## Objects and tables

  - Three tables of the **same** top height. Default `--table-z -0.220 m` (table top 22 cm below the arm-base midpoint
  at spine 0.468 m). Override per table with `--letter-table-z` / `--placement-table-z` if heights differ.
  - Pick table: a tray (1 cm thick) carrying a **cup** (8 cm tall), a **bowl containing dark beans** (4.5 cm tall, rim
  Ø 11.5 cm) with a spoon, and a **plate** as distractor. The bowl is distinguished from the plate by dark-pixel
  fraction inside its mask; the bowl grasp point is the rim point opposite the spoon handle. Right arm picks the cup,
  left arm picks the bowl. Object positions on the tray are free within the camera view and the arm workspace bounds
  in `configs/tmr_rgb20d.yaml`; no fixed positions are assumed.
  - Letter-side table (route 13 endpoint) and placement table (route 20 endpoint) must be clear where the arms set
  down; X/Y is wherever each arm is holding its object.

**Required initial layout on the tray**, seen from the robot (robot stands at the bottom edge, head camera looking
  across the tray):

  ```
                   far side of tray
     +---------------------------------------+
     |                                       |
     |   ( bowl )                            |
     |   beans +                  ( plate )  |
     |   spoon                               |
     |                            distractor |
     |   ( cup )                             |
     |                                       |
     +---------------------------------------+
                   near side of tray
       left  <--------------------->  right
     picked by                       not touched
     LEFT arm (bowl)
     RIGHT arm (cup)
                   ^ robot / head camera
  ```

  ## Lighting

  - Ordinary indoor lighting. Detection is a segmentation network, chosen because colour thresholds failed on shadows
  and on the pale-green arm bodies. Detection failure aborts the stage in place; the arm never descends on a guess.

  ## Reset between rounds

  1. Return the base to the marked start pose and heading.
  2. Put cup, bowl (with beans and spoon) and plate back on the tray on the pick table.
  3. Clear the letter-side and placement tables.
  4. Open both grippers; arms anywhere collision-free (the mission stows them).
  5. Delete `outputs/table_mission.json` or pass `--fresh-start-confirmed` — the checkpoint refuses to replay a drive
  that already happened.
  6. Confirm both impedance controllers are still active and the servo is not in fault
  (`/franka_duo/joint_servo/status`); if a reflex occurred, recover per README §8 before the next run.

### Does your policy rely on externally provided object poses?

No — our policy does its own perception end to end

### Did you train on the organizer-released trajectory dataset?

No — we did not use trajectory data

### Submission requirements

- [x] The repository is public and the pinned commit is reachable in it.
- [x] It contains a Dockerfile that builds the image named in the run commands above.
- [x] It contains a README explaining how to run the policy on the real testbed.
- [x] We have run the commands above from a clean checkout of the pinned commit, and they work.

### Acknowledgement

- [x] We understand organizers evaluate the pinned commit exactly as submitted, and will not modify our code to make it build or run.
- [x] We understand the operator has full authority to stop a run at any time, for safety or to protect the hardware.
- [x] We understand that if organizers cannot build and launch our submission from the fields above, it cannot be scored.
